spvc/test/run_spirv_cross_tests.py

```diff
@@ -45,127 +43,112 @@
     test_count += 1
-    if spvc(binary_path, tmpfile, flags) == 0 and filecmp.cmp(tmpfile, reference_path):
+    cmd = [spvc_path] + flags + ['-o', out, '--validate=vulkan1.1', inp]
+    return subprocess.call(cmd, stdout=devnull)
+
+def check_reference(result, reference):
+    global spirv_cross_dir
+    global pass_count
+    if result and filecmp.cmp(result, os.path.join(spirv_cross_dir, reference), False):
         pass_count += 1
 
-def skip_test():
-    global test_count
-    test_count += 1
+def remove_files(*filenames):
+    for i in filenames:
+        try:
+            os.remove(i)
+        except:
+            pass
+
+def test_glsl(shader, filename, optimize):
+    global spirv_cross_dir, tmpfile
+    shader_path = os.path.join(spirv_cross_dir, shader)
 
-def test_shader(binary_path, reference_path, filename, language, optimize):
-    # Note: filename is the last component of binary_path.
-    if language == 'glsl':
+    # compile shader
+    if '.asm.' in filename:
         flags = []
-        if not '.noeliminate' in filename:
-            flags.append('--remove-unused-variables')
-        if '.legacy.' in filename:
-            flags.append('--version 100 --es')
-        if '.flatten.' in filename:
-            flags.append('--flatten-ubo')
-        if '.flatten_dim.' in filename:
-            flags.append('--flatten-multidimensional-arrays')
-        if '.sso.' in filename:
-            flags.append('--separate-shader-objects')
-
-        if not '.nocompat.' in filename:
-            do_test(binary_path, reference_path, flags)
-        if '.vk.' in filename:
-            do_test(binary_path, reference_path + '.vk', flags + ['--vulkan-semantics'])
-    else:
-        # TODO: other languages
-        skip_test()
-
-def run_tests(binary_dir, source_dir):
-    '''
-    If binary_dir exists, run the tests in it, otherwise run the tests in
-    source_dir, saving the compiled form to binary_dir as we go.
-    '''
-    if os.path.isdir(binary_dir):
-        top_dir = binary_dir
-        use_source = False
+        if '.preserve.' in filename:
+            flags.append('--preserve-numeric-ids')
+        spirv_as(shader_path, tmpfile, flags)
     else:
-        if source_dir is None:
-            print('Binary directory "' + binary_dir + '"not found and no source directory given',
-                  file=sys.stderr)
-            sys.exit(1)
-        top_dir = source_dir
-        use_source = True
+        glslang_compile(shader_path, tmpfile, ['--target-env', 'vulkan1.1', '-V'])
+    if optimize and not '.noopt.' in filename and not '.invalid.' in filename:
+        spirv_opt(tmpfile, tmpfile, [])
+
+    # run spvc
+    flags = []
+    output = output_vk = ""
+    if not '.nocompat.' in filename:
+        output = tmpfile + filename
+        if spvc(tmpfile, output, flags):
+            output = ""
+    if '.vk.' in filename or '.asm.' in filename:
+        output_vk = tmpfile + 'vk' + filename
+        if spvc(tmpfile, output_vk, flags):
```

Comment:
I can see why you are tempted to put the test counting in the low level utility function.  Probably because there's so much business logic in constructing the right command lines.

I'd approach it differently.  I'd construct the elaborated list of command lines, then execute them one by one, counting at that point.  That would also give an easy way to add a -n or --dry-run option to preview exactly what's going to execute before it's executed.

That also would resolve the usability problem where: when something goes wrong, it's annoying to have to guess how to reconstruct exactly one test case for iterative fix/debug.

Reply:
It's a little more complicated to separate listing the steps and running through them.
If spvc fails you need to skip the comparison step.  The step could maybe be "spvc then compare if needed" but there's a property here I want to preserve, namely doing everything in the same order as spirv-cross.  That way I can log what both are doing and easily compare.  I had logging here temporarily and removed it - will add later as an official feature.  That also addresses the issue of knowing what went wrong (tail the log).  The logging in spirv-cross was similarly temporary and I'll try to land it there eventually.
Ideally some day we get rid of this duplication, but I'm proceeding cautiously on trying to change spirv-cross.
I think this is about as simple as possible at the moment, and that's a good thing.  I'd like to flesh out the MSL and HLSL testing a bit and see what complications that brings before deciding on another direction for this script.